File: 985-Bag of Tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(),tokens.end());
        int n = tokens.size();
        int l = 0, r = n - 1;
        int res = 0;
        int cnt = 0;

        while(l <= r){
            if (power >= tokens[l]){
                power -= tokens[l];
                cnt ++;
                l ++;
            }
            else if(l < r && cnt > 0){
                power += tokens[r];
                r--;cnt--;
            }
            else{
                break;
            }
            
            res = max(res, cnt);
        }
        return res;
    }
};

File: 985-Bag of Tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        res = 0
        tokens.sort()
        N = len(tokens)
        r = N - 1
        l = 0
        res = 0
        cnt = 0
        #print(tokens)
        while l <= r:
            if power >= tokens[l]:
                power -= tokens[l]
                l += 1
                cnt += 1
            
            elif cnt > 0 and power < tokens[l]:
                power += tokens[r]
                cnt -= 1
                r -= 1
            else:
                break # can't do anythign with the current result
            #print(power, cnt, l, r)
            res = max(res, cnt)
        
        return res
            
            